module/zfs/zfs_fm.c

```diff
@@ -931,9 +939,60 @@ zfs_post_autoreplace(spa_t *spa, vdev_t *vd)
  * open because the device was not found (fault.fs.zfs.device).
  */
 void
-zfs_post_state_change(spa_t *spa, vdev_t *vd)
+zfs_post_state_change(spa_t *spa, vdev_t *vd, uint64_t prevstate)
 {
-	zfs_post_common(spa, vd, FM_RSRC_CLASS, FM_RESOURCE_STATECHANGE);
+#ifdef _KERNEL
+	nvlist_t *aux;
+
+	/*
+	 * Add optional supplemental keys to payload
+	 */
+	aux = fm_nvlist_create(NULL);
+	if (vd && aux) {
+		char *state;
+
+		switch (vd->vdev_state) {
+		default:
+			state = "UNKNOWN";
+			break;
+		case VDEV_STATE_CLOSED:
+		case VDEV_STATE_OFFLINE:
+			state = "OFFLINE";
+			break;
+		case VDEV_STATE_REMOVED:
+			state = "REMOVED";
+			break;
+		case VDEV_STATE_CANT_OPEN:
+			state = "UNAVAIL";
+			break;
+		case VDEV_STATE_FAULTED:
+			state = "FAULTED";
+			break;
+		case VDEV_STATE_DEGRADED:
+			state = "DEGRADED";
+			break;
+		case VDEV_STATE_HEALTHY:
+			state = "ONLINE";
+			break;
+		}
+		if (vd->vdev_physpath) {
+			(void) nvlist_add_string(aux,
+			    FM_EREPORT_PAYLOAD_ZFS_VDEV_PHYSPATH,
+			    vd->vdev_physpath);
+		}
+		(void) nvlist_add_uint64(aux,
+		    FM_EREPORT_PAYLOAD_ZFS_VDEV_PREVSTATE, prevstate);
+
+		(void) nvlist_add_string(aux,
+		    FM_EREPORT_PAYLOAD_ZFS_VDEV_STATE_STR, state);
+	}
+
+	zfs_post_common(spa, vd, FM_RSRC_CLASS, FM_RESOURCE_STATECHANGE,
```

Comment:
I prefer to keep strings close to the enum, close enough so that an update of one is more likely to see an update in the other.

That said, I'd also rather see a table than a case.